File: 0235_med_hot_lowest-common-ancestor-of-a-binary-search-tree/solution.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        r = root

        while not (p.val <= r.val <= q.val or q.val <= r.val <= p.val):
        
            if p.val < r.val and q.val < r.val:
                r = r.left
            if p.val > r.val and q.val > r.val:
                r = r.right

        return r
    

class Solution2:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        if not root:
            return None
        
        if q.val < root.val and p.val < root.val:
            return self.lowestCommonAncestor(root.left, p, q)
        elif q.val > root.val and p.val > root.val:
            return self.lowestCommonAncestor(root.right, p, q)
        else:
            return root

File: 0235_med_hot_lowest-common-ancestor-of-a-binary-search-tree/solution.java
/**
 * Definition for a binary tree node.
 */

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode(int x) { val = x; }
}

 

 class Solution {
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        if (root == null) {
            return null;
        }

        if(p.val < root.val && q.val < root.val) {
            return lowestCommonAncestor(root.left, p, q);
        } else if (p.val > root.val && q.val > root.val) {
            return lowestCommonAncestor(root.right, p, q);
        } else {
            return root;
        }
    }
}


class Solution2 {
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        if (root == null) {
            return null;
        }

        TreeNode ptr = root;

        while(ptr != null) {
            if(p.val < ptr.val && q.val < ptr.val) {
                ptr = ptr.left;
            } else if (p.val > ptr.val && q.val > ptr.val) {
                ptr = ptr.right;
            } else {
                // found
                break;
            }
        }

        return ptr;
    }
}